### DISCUSSION TITLE
How to monitor TUN/TAP interface

### DISCUSSION THREAD

**Discussion by ssrlive** [libuv/libuv#3502] [Q&A]
How to monitor the arrival of data on the TUN/TAP interface? Currently, I use a timer to read in a loop, but the performance is very poor. Is there any better solution?

https://github.com/ssrlive/toyvpn/blob/232e859f19152a6663b0a5764dcadde433cc6140/server/linux/ToyVpnServer2.c#L328-L334

**Comment by bnoordhuis** [comment]
I think you should be able to create a `uv_poll_t` out of `tun_fd` and get notified of incoming data that way.

**Reply by ssrlive** [reply]
The `uv_poll_t` component has been applied, but this triggers another problem: before, the UDP listening routine behaved normally, but now, the data received is very strange, often receiving huge packets, and often fail when writing huge packets to the TUN interface. How to fix this side effect?

This strange behavior is causing the program to not work correctly at all, not just a performance issue anymore.

https://github.com/ssrlive/toyvpn/blob/b31e4cf6f72eb138a878e894504e0d5e44096d61/server/linux/ToyVpnServer2.c#L529-L562

```
Write size: 56
Write size: 56
Write size: 56
Write size: 56
Write size: 32766
Write (size 32766) error: invalid argument
Write size: 32766
Write (size 32766) error: invalid argument
Write size: 32766
Write (size 32766) error: invalid argument
Write size: 56
Write size: 56
```

**Reply by bnoordhuis** [reply]
Can you try strace-ing it?

**Reply by ssrlive** [reply]
No, In my experience, these functions can't be used in wrong way. So this is very strange.

**Reply by bnoordhuis** [reply]
I mean, can you run strace on your program? I'd like to see what system calls are failing.

**Reply by ssrlive** [reply]
Here is the log file of strace. I don't know which function cause the issue.

[a.zip](https://github.com/libuv/libuv/files/8185621/a.zip)

**Reply by bnoordhuis** [reply]
Sorry, missed your reply. I had a look but I don't see any relevant failing system calls. Can you post a trace where your program prints that error?
